debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/SnapshotMySQLUtility.java

```diff
@@ -0,0 +1,600 @@
+        this.metrics = new SnapshotReaderMetrics(context.clock());
+        this.mysql = context.jdbc();
+    }
+
+    /**
+     * Set whether to block other transactions as minimally as possible by releasing the read lock
+     * as early as possible (when buffer is first called). Although the snapshot process should
+     * obtain a consistent snapshot even when releasing the lock as early as possible, it may be
+     * desirable to explicitly hold onto the read lock until execution completes. In such cases,
+     * holding onto the lock will prevent all updates to the database during the snapshot process.
+     *
+     * @param minimalBlocking {@code true} if the lock is to be released as early as possible, or
+     *                        {@code false} if the lock is to be held for the entire transaction.
+     * @return this object for method chaining; never null
+     */
+    public void useMinimalBlocking(boolean minimalBlocking) {
+        this.minimalBlocking = minimalBlocking;
+    }
+
+    /**
+     * Begin a transaction and read the binlog position. Required before calling
+     * {@link #bufferLockedTables(SnapshotServerReader.RecordRecorder, BufferedBlockingConsumer)} or
+     * {@link #bufferLockedTables(List, SnapshotServerReader.RecordRecorder, BufferedBlockingConsumer)}.
+     * @throws SQLException
+     */
+    public void beginTransaction() throws SQLException {
+        // Set the transaction isolation level to REPEATABLE READ. This is the default, but the default can be changed
+        // which is why we explicitly set it here.
+        //
+        // With REPEATABLE READ, all SELECT queries within the scope of a transaction (which we don't yet have) will read
+        // from the same MVCC snapshot. Thus each plain (non-locking) SELECT statements within the same transaction are
+        // consistent also with respect to each other.
+        //
+        // See: https://dev.mysql.com/doc/refman/5.7/en/set-transaction.html
+        // See: https://dev.mysql.com/doc/refman/5.7/en/innodb-transaction-isolation-levels.html
+        // See: https://dev.mysql.com/doc/refman/5.7/en/innodb-consistent-read.html
+        logger.info("\tdisabling autocommit and enabling repeatable read transactions");
+        mysql.setAutoCommit(false);
+        sql.set("SET TRANSACTION ISOLATION LEVEL REPEATABLE READ");
+        mysql.execute(sql.get());
+        // Start a transaction
+        // See http://dev.mysql.com/doc/refman/5.7/en/commit.html
+        sql.set("START TRANSACTION WITH CONSISTENT SNAPSHOT");
+        mysql.execute(sql.get());
+        activeTransaction = true;
+
+        readBinlogPosition();
+    }
+
+    /**
+     * Read and internally store all avaliable databases, and all the tables that match the config.
+     *
+     * @throws SQLException
+     * @see {@link #getTableIds()}, {@link #getTableIdsByDbName()}, {@link #getReadableDatabaseNames()}
+     */
+    public void readTableList() throws SQLException {
```

Comment:
A general remark on this class and is methods is that it seems to be based a lot on side effects. So e.g. this method seems to alter the state of this class instead of returning the list of tables. I feel this makes it tough to grok the overarching picture.